sympy/matrices/common.py

```diff
@@ -753,6 +755,40 @@ def size(m):
                         diag_entries[(i + row_pos, j + col_pos)] = m[i, j]
                 row_pos += m.rows
                 col_pos += m.cols
+            elif(isinstance(m,dict)):
+                # in this case we're a dict
+                for key,value in m.items():
+                    r_p = 0
+                    c_p = 0
+                    r_p = 0 if key>0 else -(key)
+                    c_p = key if key>0 else 0
+                    func_arg = 0
+                    while (r_p<diag_rows and c_p<diag_cols):
+                        from inspect import isfunction, getargspec
+
+                        if(isinstance(value,list)):
+                            if(len(value)!=min(diag_rows-r_p,diag_cols-c_p)):
+                                raise ValueError("The size of list provided should match the diagonal size")
+                            for i in range(len(value)):
+                                diag_entries[(r_p,c_p)] = value[i]
+                                r_p += 1
+                                c_p += 1
+                        elif(isfunction(value)):
+                            num_args = len(getargspec(value).args)
+                            if(num_args==2):
+                                diag_entries[(r_p,c_p)] = value(r_p,c_p)
+                                r_p+=1
+                                c_p+=1
+                            if(num_args==1):
```

Comment:
if->elif and then the two increments (e.g. r_p += 1) can be moved to after the if-elif block